### ISSUE TITLE
Toil doesn't work with GridEngine on PODK

### ISSUE THREAD

**Issue by melissacline** [BD2KGenomics/toil#572]
I was trying to run the UNC RNASeq pipeline on podk, with gridEngine.  It failed almost immediately, with the following output in the error log:

-bash-4.1$ cat launch_ispy.11.23.15.errout
No handlers could be found for logger "toil.resource"
INFO:toil.lib.bioio:Logging set at level: INFO
INFO:toil.common:Using the grid engine machine batch system
INFO:toil.jobStores.fileJobStore:Jobstore directory is: /scratch/cline/toil_mnt/jstore
INFO:toil.common:Written the environment for the jobs to the environment file
WARNING:toil.resource:Can't globalize module ModuleDescriptor(dirPath='/pod/pstore/projects/ispy2/scripts', name='rnaseq_unc_pipeline', extension='.py').
INFO:toil.leader:Checked batch system has no running jobs and no updated jobs
INFO:toil.leader:Found 1 jobs to start and 0 jobs with successors to run
INFO:toil.leader:Starting the main loop
WARNING:toil.leader:Batch system is reporting that the jobWrapper with batch system ID: 0 and jobWrapper store ID: 3/D/job3rR7C_ failed with exit value 1
WARNING:toil.leader:No log file is present, despite jobWrapper failing: 3/D/job3rR7C_
WARNING:toil.jobWrapper:Due to failure we are reducing the remaining retry count of job 3/D/job3rR7C_ to 0
WARNING:toil.leader:Job: 3/D/job3rR7C_ is completely failed
INFO:toil.leader:Only failed jobs and their dependents (1 total) are remaining, so exiting.
INFO:toil.leader:Finished the main loop
INFO:toil.leader:Waiting for stats and logging collator process to finish
INFO:toil.leader:Stats/logging finished collating in 0.169120073318 seconds
Traceback (most recent call last):
  File "rnaseq_unc_pipeline.py", line 838, in <module>
    main()
  File "rnaseq_unc_pipeline.py", line 835, in main
    Job.Runner.startToil(Job.wrapJobFn(download_shared_files, inputs), args)
  File "/pod/home/cline/lib/python2.7/site-packages/toil/job.py", line 372, in startToil
    return mainLoop(config, batchSystem, jobStore, rootJob)
  File "/pod/home/cline/lib/python2.7/site-packages/toil/leader.py", line 505, in mainLoop
    raise FailedJobsException( config.jobStore, totalFailedJobs )
toil.leader.FailedJobsException: The job store '/scratch/cline/toil_mnt/jstore' contains 1 failed jobs

**Comment by jvivian** [comment]
I've run into a reproducible issue running the rna-seq pipeline on PODK.   It handles several jobs, then fails with:

```
DEBUG:toil.batchSystems.gridengine:List of running jobs: set([8])
DEBUG:toil.batchSystems.gridengine:Exit Status: '0'
DEBUG:toil.batchSystems.gridengine:List of running jobs: set([])
DEBUG:toil.batchSystems.gridengine:No activity, sleeping for 1s
DEBUG:toil.batchSystems.gridengine:UpdatedJobsQueue Item: (8, 0)
DEBUG:toil.leader:Batch system is reporting that the jobWrapper with batch system ID: 8 and jobWrapper store ID: 9/S/jobXCy1ZM ended successfully
DEBUG:toil.leader:Added jobWrapper: 9/S/jobXCy1ZM to active jobs
DEBUG:toil.leader:Built the jobs list, currently have 1 jobs to update and 0 jobs issued
DEBUG:toil.leader:Job: 9/S/jobXCy1ZM has 2 successors to schedule
DEBUG:toil.batchSystems.gridengine:Issued the job command: /pod/home/jvivian/venv/bin/_toil_worker /pod/pstore/jvivian/jstore V/0/jobiDtxOg with job id: 9
DEBUG:toil.leader:Issued job with job store ID: V/0/jobiDtxOg and job batch system ID: 9 and cores: 32, disk: 161061273600, and memory: 42949672960
DEBUG:toil.batchSystems.gridengine:Issued the job command: /pod/home/jvivian/venv/bin/_toil_worker /pod/pstore/jvivian/jstore k/U/jobaUFbmq with job id: 10
DEBUG:toil.leader:Issued job with job store ID: k/U/jobaUFbmq and job batch system ID: 10 and cores: 32, disk: 107374182400, and memory: 2147483648
DEBUG:toil.batchSystems.gridengine:Running ['qsub', '-b', 'y', '-terse', '-j', 'y', '-cwd', '-o', '/pod/home/jvivian/39501.9.out', '-e', '/pod/home/jvivian/39501.9.err', '-N', 'toil_job_9', '-hard', '-l', 'vf=41943040K,h_vmem=41943040K', '-pe', 'shm', '32', '/pod/home/jvivian/venv/bin/_toil_worker /pod/pstore/jvivian/jstore V/0/jobiDtxOg']
Unable to run job: job rejected: the requested parallel environment "shm" does not exist.
Exiting.
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/pod/home/jvivian/python-2.7.3/lib/python2.7/threading.py", line 551, in __bootstrap_inner
    self.run()
  File "/pod/home/jvivian/venv/lib/python2.7/site-packages/toil/batchSystems/gridengine.py", line 186, in run
    activity |= self.createJobs(newJob)
  File "/pod/home/jvivian/venv/lib/python2.7/site-packages/toil/batchSystems/gridengine.py", line 158, in createJobs
    sgeJobID = self.qsub(qsubline)
  File "/pod/home/jvivian/venv/lib/python2.7/site-packages/toil/batchSystems/gridengine.py", line 215, in qsub
    result = int(process.stdout.readline().strip().split('.')[0])
ValueError: invalid literal for int() with base 10: '
```

which comes from this line:
https://github.com/BD2KGenomics/toil/blob/master/src/toil/batchSystems/gridengine.py#L208

I tried a few different environments, but changing `shm` to `mpi` fixed it on podk.  I've looked around but I'm unsure of the most appropriate way to check whether MPI, OpenMPI, RMI, SHM, etc... is the system's given parallel environment.  @hannes-ucsc , is it worth testing a PR to see if an `mpi` environment would build on jenkins? I'm not sure if that's where TOIL_GRIDENGINE_PE comes into play?  If you have an idea of what we could implement let me know and I can start working on a PR.

**Comment by jvivian** [comment]
I tried again on Toil 3.1.1 and this was the error I got using gridEngine.  I was able to ssh into a single node (ssh jvivian@podk-1-2) and run it in single machine mode. 

```
DEBUG:toil.batchSystems.gridengine:List of running jobs: set([0])
DEBUG:toil.batchSystems.gridengine:Exit Status: '1'
DEBUG:toil.batchSystems.gridengine:List of running jobs: set([])
DEBUG:toil.batchSystems.gridengine:No activity, sleeping for 1s
DEBUG:toil.batchSystems.gridengine:UpdatedJobsQueue Item: (0, 1)
WARNING:toil.leader:Batch system is reporting that the jobWrapper with batch system ID: 0 and jobWrapper store ID: w/8/jobPD241g failed with exit value 1
WARNING:toil.leader:No log file is present, despite jobWrapper failing: w/8/jobPD241g
WARNING:toil.jobWrapper:Due to failure we are reducing the remaining retry count of job w/8/jobPD241g to 0
DEBUG:toil.leader:Added jobWrapper: w/8/jobPD241g to active jobs
DEBUG:toil.leader:Built the jobs list, currently have 1 jobs to update and 0 jobs issued
WARNING:toil.leader:Job: w/8/jobPD241g is completely failed
INFO:toil.leader:Only failed jobs and their dependents (1 total) are remaining, so exiting.
INFO:toil.leader:Finished the main loop
INFO:toil.leader:Waiting for stats and logging collator process to finish ...
INFO:toil.leader:... finished collating stats and logs. Took 0.432924985886 seconds
DEBUG:toil.common:Shutting down batch system
DEBUG:toil.batchSystems.gridengine:Received queue sentinel.
Traceback (most recent call last):
  File "rnaseq_cgl_pipeline.py", line 912, in <module>
    main()
  File "rnaseq_cgl_pipeline.py", line 908, in main
    Job.Runner.startToil(Job.wrapJobFn(parse_input_samples, inputs), args)
  File "/pod/home/jvivian/python-2.7.3/lib/python2.7/site-packages/toil/job.py", line 437, in startToil
    return mainLoop(config, batchSystem, jobStore, rootJob)
  File "/pod/home/jvivian/python-2.7.3/lib/python2.7/site-packages/toil/leader.py", line 507, in mainLoop
    rootJobReturnValue = cPickle.load(fH)
EOFError
```

**Comment by melissacline** [comment]
So does this mean that Toil and GridEngine seem to be working together, and
the error was with the input file?

**Comment by hannes-ucsc** [comment]
@jvivian, this may mean progress. The key is that the job failed without producing output. So it failed early, possibly before being able to launch the worker process. Unfortunately, bot stderr and out are sent to /dev/null in prepareQsub(). Maybe if we removed that redirection we would get the $jobID.OUT and .ERR files with more info on what's going on?

**Comment by hannes-ucsc** [comment]
Output issues will be addressed by #620.
